timeseries/src/autogluon/timeseries/models/chronos/utils.py

```diff
@@ -1,11 +1,177 @@
+
+    def __init__(
+        self,
+        tokenizer: ChronosTokenizer,
+        target_df: TimeSeriesDataFrame,
+        freq: str,
+        target_column: str = "target",
+        context_length: int = 512,
+        prediction_length: int = 64,
+        mode: Literal["training", "validation"] = "training",
+    ) -> None:
+        super().__init__()
+
+        assert mode in ("training", "validation")
+
+        self.gluonts_dataset = SimpleGluonTSDataset(target_df=target_df, freq=freq, target_column=target_column)
+        self.tokenizer = tokenizer
+        self.context_length = context_length
+        self.prediction_length = prediction_length
+        self.mode = mode
+
+    def preprocess_entry(self, entry: dict) -> dict:
+        entry = {f: entry[f] for f in [FieldName.START, FieldName.TARGET]}
+        entry[FieldName.TARGET] = np.asarray(entry[FieldName.TARGET], dtype=np.float32)
+
+        return entry
+
+    def _create_instance_splitter(self, mode: str):
+        instance_sampler = {
+            "training": ExpectedNumInstanceSampler(num_instances=1.0, min_future=self.prediction_length),
+            "validation": ValidationSplitSampler(min_future=self.prediction_length),
+        }[mode]
+
+        return InstanceSplitter(
+            target_field=FieldName.TARGET,
+            is_pad_field=FieldName.IS_PAD,
+            start_field=FieldName.START,
+            forecast_start_field=FieldName.FORECAST_START,
+            instance_sampler=instance_sampler,
+            past_length=self.context_length,
+            future_length=self.prediction_length,
+            dummy_value=np.nan,
+        )
+
+    def _create_training_data(self, data):
+        data = Cyclic(data)
+        split_transform = self._create_instance_splitter("training")
+        data = split_transform.apply(data, is_train=True)
+        return data
+
+    def _create_validation_data(self, data):
+        data = self._create_instance_splitter("validation").apply(data, is_train=False)
+        return data
+
+    def to_hf_format(self, entry: dict) -> dict:
+        past_target = torch.tensor(entry[f"past_{FieldName.TARGET}"]).unsqueeze(0)
+        input_ids, attention_mask, scale = self.tokenizer.context_input_transform(past_target)
+        future_target = torch.tensor(entry[f"future_{FieldName.TARGET}"]).unsqueeze(0)
+        labels, labels_mask = self.tokenizer.label_input_transform(future_target, scale)
+        labels[labels_mask == 0] = -100
+
+        return {
+            "input_ids": input_ids.squeeze(0),
+            "attention_mask": attention_mask.squeeze(0),
+            "labels": labels.squeeze(0),
+        }
+
+    def __iter__(self) -> Iterator:
+        preprocessed_dataset = Map(self.preprocess_entry, self.gluonts_dataset)
+
+        if self.mode == "training":
+            iterable = self._create_training_data(preprocessed_dataset)
+        elif self.mode == "validation":
+            iterable = self._create_validation_data(preprocessed_dataset)
```

Comment:
Is there a way to limit the # of validation batches (similar to https://lightning.ai/docs/pytorch/stable/common/trainer.html#limit-val-batches)? Otherwise we will just spend most of the time computing the validation loss for datasets with a large number of items.

Reply:
Alternatively, we can just randomly subsample the items when creating the `tokenizer_val_dataset` above.

Reply:
There's no such option in `transformers` it seems, so I did it manually.